pkg/services/accesscontrol/resourcepermissions/api.go

```diff
@@ -30,18 +30,45 @@ func newApi(ac accesscontrol.AccessControl, router routing.RouteRegister, manage
 	return &api{ac, router, manager, permissions}
 }
 
+func (a *api) getEvaluators(actionRead, actionWrite, scope string) (read, write accesscontrol.Evaluator) {
+	if a.service.options.InheritedScopesSolver == nil || a.service.options.InheritedScopePrefixes == nil {
+		read = accesscontrol.EvalPermission(actionRead, scope)
+		write = accesscontrol.EvalPermission(actionWrite, scope)
+	} else {
+		// Add inherited scopes to the evaluators protecting the endpoint.
+		// Scopes in the request context parameters are to be added by solveInheritedScopes.
+		// If a user got actionRead on any of the inherited scopes, they will be granted access to the endpoint.
+		// Ex: a user inherits dashboards:read from the containing folder (folders:uid:BCeknZL7k)
+		inheritedRead := []accesscontrol.Evaluator{accesscontrol.EvalPermission(actionRead, scope)}
+		inheritedWrite := []accesscontrol.Evaluator{accesscontrol.EvalPermission(actionWrite, scope)}
+		for _, scopePrefix := range a.service.options.InheritedScopePrefixes {
+			inheritedRead = append(inheritedRead,
+				accesscontrol.EvalPermission(actionRead, accesscontrol.Parameter(scopePrefix)))
+			inheritedWrite = append(inheritedWrite,
+				accesscontrol.EvalPermission(actionWrite, accesscontrol.Parameter(scopePrefix)))
+		}
+
+		read = accesscontrol.EvalAny(inheritedRead...)
+		write = accesscontrol.EvalAny(inheritedWrite...)
+	}
+	return
+}
+
 func (a *api) registerEndpoints() {
 	auth := middleware.Middleware(a.ac)
 	uidSolver := solveUID(a.service.options.UidSolver)
+	inheritanceSolver := solveInheritedScopes(a.service.options.InheritedScopesSolver)
 	disable := middleware.Disable(a.ac.IsDisabled())
 	a.router.Group(fmt.Sprintf("/api/access-control/%s", a.service.options.Resource), func(r routing.RouteRegister) {
+		actionRead := fmt.Sprintf("%s.permissions:read", a.service.options.Resource)
+		actionWrite := fmt.Sprintf("%s.permissions:write", a.service.options.Resource)
 		scope := accesscontrol.Scope(a.service.options.Resource, a.service.options.ResourceAttribute, accesscontrol.Parameter(":resourceID"))
-		actionWrite, actionRead := fmt.Sprintf("%s.permissions:write", a.service.options.Resource), fmt.Sprintf("%s.permissions:read", a.service.options.Resource)
+		readEvaluator, writeEvaluator := a.getEvaluators(actionRead, actionWrite, scope)
```

Comment:
In my option, they are different and I'd rather have them named differently to avoid the confusion.
* `evaluator` is the name given to the structure able to evaluate a user has been granted a set of permissions, it is used by the accesscontrol authorize middleware.
* here `solver` is a middleware computing something for later usage. The `uidSolver` defined here replaces the value stored in the reqContext ":resourceID" from UID to ID instead. The `inheritanceSolver` adds inherited scopes (folder scopes in this specific case) to the reqContext.

Reply:
@gamab 
if this is the case then i get that distinction!
do we use this pattern elsewhere? maybe we do actually if im not mistaken. 🤔